### ISSUE TITLE
Normal mapping tangents and bitangents

### ISSUE THREAD

**Issue by adnanademovic** [sotrh/learn-wgpu#258]
First of all, great tutorial! There's a lot of API to sift through, and the introduction reduced my learning time from probably months to 3 days.

I have some issues that I noticed in tangents/bitangents calculation:
https://sotrh.github.io/learn-wgpu/intermediate/tutorial11-normals/#the-tangent-and-the-bitangent

First of all, this construct creates a vector of `[0, 1, 2, ..., indices.len() - 1]`, when I assume you want a vector of zeros:

```rust
// tutorial
let mut triangles_included = (0..vertices.len()).collect::<Vec<_>>();
// correct
mut triangles_included = vec![0; vertices.len()];
```

I would have just done a PR, but there's a bigger issue that makes it irrelevant. You are doing:
```rust
vertex.tangent = (cgmath::Vector3::from(vertex.tangent) * denom).normalize().into();
```

The denom is not needed - it's always non-negative, so it won't affect the vector after it is normalized. Thus, the whole `triangles_included` process is unnecessary.

**Comment by sotrh** [comment]
My understanding was that the `demon` part was needed to average the contributions of the different vectors. Are you saying the that normalize is all that's needed for that?

**Comment by adnanademovic** [comment]
No matter the length of the vector, after `normalize()` is called the length becomes 1.

If you didn't have the normalize call, then denom would be doing that job. But in that case, the triangle counting vector should be initialized to zeros.

**Comment by wojtekpil** [comment]
I agree, denom is not needed here.  Dividing or multiplying vectors by scalars is meaningless if we normalize them afterwards. You can just omit it (whole `triangles_included` process), result should be the same:
```rust
vertex.tangent = cgmath::Vector3::from(vertex.tangent).normalize().into();
```
I think that if we take a sum of normalized vectors (tangents) and divide it by its count we basically doing normalization manually. And yes, this line is actually making things worse:
```rust
let mut triangles_included = (0..vertices.len()).collect::<Vec<_>>();
```
because without `.normalize()` call the logic is not correct, as the counts are wrong. Moreover with bigger meshes it may introduce some numerical precision problems, as we could divide normalized vectors by very large numbers.

**Comment by sotrh** [comment]
I have a fix for this in the `web2` branch. I'm not quite able to merge it though as one of my shaders is broken in webGL.

**Comment by sotrh** [comment]
Closed by #334
